### ISSUE TITLE
Safari doesn't like normalize() (aka Safari is the new IE)

### ISSUE THREAD

**Issue by bbenezech** [odysseyscience/react-s3-uploader#44]
SAFARI Version 9.0.1 (11601.2.7.2), MacOSX El President :gift: 

(I'm up to date)

You'll get errors such as:
```
TypeError: file.name.replace(/\s+/g, "_").normalize is not a function. (In 'file.name.replace(/\s+/g, "_").normalize()', 'file.name.replace(/\s+/g, "_").normalize' is undefined)
executeOnSignedUrl
uploadFile
send
onAttachmentAdd
dispatchEvent
triggerEvent
notify
compositionDidAddAttachment
refreshAttachments
setDocument
insertText
insertAttachment
insertFile
insertFile
onFileSelect
```

Happens here:
https://github.com/odysseyscience/react-s3-uploader/blob/0f1b32979cc35d7e5d8669c1d13caaaeae3a27f0/s3upload.js#L65
https://github.com/odysseyscience/react-s3-uploader/blob/0f1b32979cc35d7e5d8669c1d13caaaeae3a27f0/s3upload.js#L136

Resources:
http://stackoverflow.com/questions/27235902/javascript-normalize-causing-error-in-safari

Proposed fix:
URL encoding the shit could do it.

**Comment by bbenezech** [comment]
latinize lib is a bit crappy if it can't even deal with endpoint normalization itself (no offense to its maintainer).
I removed all sanitizing (normalize/latinize), I had no issue with safari/firefox/chrome OSX with S3 Dublin.
Tested with this name. `ééé[][]--0-9-087896,.,;';p[p[]-30nov2015_3`
Why do we need UTF-8 to be removed?

**Comment by seanadkinson** [comment]
Thanks for reporting this.

The reporter of issue #32 had experienced issues with accentuated filenames, and there is a discussion and some links in that thread if you'd like to see the history.  I think it boiled down to the fact that URLs should only have ASCII characters, so work was down to normalize filenames accordingly.

If `encodeURIComponent` works just as well, I'm fine with that as a solution instead of depending on these other libraries.  Send a PR if you want me to take a look.

**Comment by bbenezech** [comment]
I went down the rabbit hole.

AFAIK the only moment we set any header with the file name in it, is when user configures a specific Content-Disposition header.
@rogerso added Content-Disposition to the uploading part. I can't find any context as why. 
Does it make sense when uploading to a server, since you would normally use it to force a download popup client side, which means setting it server side when answering a download request?

latinize is definitely a hack and should be removed ASAP. That part I'm certain.
Do we need to encodeURI components? My instinct tells me no. You would use it as a protection to not mess up with your URL parsing server-side, but since Amazon S3 handles ?/& with grace in upload names (I tested it), this seems not necessary.

tldr; do we need Content-Disposition header setting?

**Comment by seanadkinson** [comment]
`Content-Disposition` is nice so that you can, for example, specify `inline` and have PDFs open in-browser by default.  Specifying `filename` makes it so that it downloads with the same name as you uploaded, but isn't necessarily stored on the server with that `filename` (unless I'm not mistaken).  Both seem like valid features, although I'd say the second one is more sugar than functionality, so is probably expendable.  If we can set `Content-Disposition` without `filename`, and that gets rid of all the problems, that would be a fine solution.

Another option is to just change any `str.normalize()` to `unorm.ufc(str)`, since that is what the polyfill does anyway.  Your root issue was that `"normalize() is not a function"`, correct?

I have some other projects going on, but I'll gladly look at PRs and get a release pushed if it makes sense to.  Thanks again.

**Comment by bbenezech** [comment]
I still don't understand why someone would set `Content-Disposition` when sending to a _server_. This sounds nuts.

**Comment by seanadkinson** [comment]
Not intuitive, I agree, and I didn't know about this feature until the PR was opened.  Basically it seems like you can send S3 headers that you want it to use when the file is downloaded.  So even though `Content-Disposition` doesn't make sense on a request, it is actually given for use in the response, is my understanding.

Let me put up a branch real quick with the super-minor change of `str.normalize()` to `unorm.nfc()`

**Comment by seanadkinson** [comment]
Alright, try using the `fix-normalize` branch, and let me know if it works for you.  Seems like the validity of putting the `filename` in the header is a different discussion.

**Comment by bbenezech** [comment]
Ok I get it now, thank you for taking the time to explain.

There is some background about this here: http://www.bennadel.com/blog/2696-overriding-content-type-and-content-disposition-headers-in-amazon-s3-pre-signed-urls.htm

He chose to URL encode the file name fragment, which makes sense.

Stricto-sensus, you can close the issue while merging `fix-normalize`.

However, I think all this can be simplified a lot:

the `Content-Disposition` header can be passed as a complete header to your library, you can document it as such, the specific implementation is over-specific. There are tons of other headers you could generate the same way and would have to maintain with this kind of issue. It smells like 'ad-hoc' API, which kills kittens and maintainers alike, as you probably know. 

With `Content-Disposition` out of the way, you can remove all file name sanitizing, (which should not exist anyway in the current implementation for the URL part).

**Comment by seanadkinson** [comment]
Yes, you are correct - we shouldn't have the `Content-Disposition` handling hard-coded, and users can provide that header if they wish.  That would be more scalable and maintainable, and would get rid of these dependencies.  That would be a breaking change though, so I'll wait for a bigger update to rip that out.

I'll merge and push a point release for this `normalize` fix though, thanks.

**Comment by seanadkinson** [comment]
Pushed `v1.2.1` to npm.
